planner/core/rule_column_pruning.go

```diff
@@ -265,3 +264,30 @@ func (p *LogicalLock) PruneColumns(parentUsedCols []*expression.Column) {
 		p.children[0].PruneColumns(parentUsedCols)
 	}
 }
+
+// PruneColumns implements LogicalPlan interface.
+func (p *LogicalWindow) PruneColumns(parentUsedCols []*expression.Column) {
+	windowColumn := p.GetWindowResultColumn()
+	len := 0
+	for _, col := range parentUsedCols {
+		if !windowColumn.Equal(nil, col) {
+			parentUsedCols[len] = col
+			len++
+		}
+	}
+	parentUsedCols = parentUsedCols[:len]
+	parentUsedCols = p.extractUsedCols(parentUsedCols)
+	p.children[0].PruneColumns(parentUsedCols)
+	p.SetSchema(p.children[0].Schema().Clone())
+	p.Schema().Append(windowColumn)
+}
+
+func (p *LogicalWindow) extractUsedCols(parentUsedCols []*expression.Column) []*expression.Column {
+	for _, arg := range p.WindowFuncDesc.Args {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(arg)...)
+	}
+	for _, by := range p.ByItems {
+		parentUsedCols = append(parentUsedCols, by.Col)
+	}
+	return parentUsedCols
+}
```

Comment:
Shall we implement other logical optimizations for `LogicalWindow` as well? e.g, `LogicalWindow` does things like projection, so it may impact behavior of projection elimination? another logical rule I can think of is predicate push down, for queries like `select * from (select a, row_number() over (partition by b) from t) t1 where t1.a > 1`, can we push the `t1.a > 1` filter over `LogicalWindow`? seems it may impact the result of window function?

Reply:
Yes, the projection elimination needs to be considered. And I don't think we can push down the filters, it affects the final result.

Reply:
Added.